staging/src/k8s.io/apiserver/pkg/endpoints/handlers/apply.go

```diff
@@ -0,0 +1,180 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package handlers
+
+import (
+	"fmt"
+
+	"github.com/ghodss/yaml"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/apply"
+	"k8s.io/apimachinery/pkg/apply/parse"
+	"k8s.io/apimachinery/pkg/apply/strategy"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/json"
+	"k8s.io/kube-openapi/pkg/util/proto"
+)
+
+type applyPatcher struct {
+	*patcher
+
+	model proto.Schema
+}
+
+// TODO(apelisse): workflowId needs to be passed as a query
+// param/header, and a better defaulting needs to be defined too.
+const workflowId = "default"
+
+func (p *applyPatcher) convertCurrentVersion(obj runtime.Object) (map[string]interface{}, error) {
+	vo, err := p.unsafeConvertor.ConvertToVersion(obj, p.hubGroupVersion)
+	if err != nil {
+		return nil, err
+	}
+	return runtime.DefaultUnstructuredConverter.ToUnstructured(vo)
+}
+
+func (p *applyPatcher) extractLastIntent(obj runtime.Object, workflow string) (map[string]interface{}, error) {
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't get accessor: %v", err)
+	}
+	last := make(map[string]interface{})
+	// TODO: use the managedFields correctly
+	if _, ok := accessor.GetManagedFields()[workflow]; ok {
+		if err := json.Unmarshal([]byte(accessor.GetManagedFields()[workflow].APIVersion), &last); err != nil {
```

Comment:
We may need to increase the object size limit in etcd to allow for this extra data